packages/tailwind-typography/src/index.ts

```diff
@@ -0,0 +1,205 @@
+import plugin from 'tailwindcss/plugin'
+
+function isPlainObject(x: unknown): x is Record<string, unknown> {
+  return (
+    typeof x === 'object' &&
+    x?.constructor === Object &&
+    Object.getPrototypeOf(x) === Object.prototype
+  )
+}
+
+export const buildTypographyVariant = (
+  fontFamily: string,
+  fontWeight: number,
+  fontSize: number,
+  lineHeight: number,
+  letterSpacing: number,
+  other?: Record<string, unknown>,
+) => ({
+  fontFamily,
+  fontWeight,
+  fontSize,
+  // Unit-less following http://meyerweb.com/eric/thoughts/2006/02/08/unitless-line-heights/
+  lineHeight,
+  letterSpacing: `${letterSpacing}em`,
+  ...other,
+})
+
+export function fluidValue(
+  minFontSize: number,
+  maxFontSize: number,
+  minViewport: number,
+  maxViewport: number,
+  unit = 'px',
+  clampMin = false,
+  clampMax = false,
+) {
+  // Set up the fluid formula.
+  const deltaFontSize = maxFontSize - minFontSize
+  const deltaViewport = maxViewport - minViewport
+  const fluidFontSize = `calc(${minFontSize}${unit} + ${deltaFontSize} * ((100vw - ${minViewport}${unit}) / ${deltaViewport}))`
+
+  // Clamp fontSize values if enabled.
+  if (clampMin && clampMax) {
+    return `clamp(${minFontSize}${unit}, ${fluidFontSize}, ${maxFontSize}${unit})`
+  }
+  if (clampMin) {
+    return `max(${minFontSize}${unit}, ${fluidFontSize})`
+  }
+  if (clampMax) {
+    return `min(${minFontSize}${unit}, ${fluidFontSize})`
+  }
+  return fluidFontSize
+}
+
+export interface TypographyOptionsVariant {
+  fontSize: number
+  [key: string]: string | number
+}
+
+export type TypographyOptionsVariantMap<TKeys extends readonly [string, ...string[]]> = {
+  clamp?: boolean
+  clampMax?: boolean
+  clampMin?: boolean
+  fluid?: boolean
+} & {
+  [key in TKeys[number]]?: TypographyOptionsVariant
+} & {
+  [key in TKeys[0]]-?: TypographyOptionsVariant
+}
```

Comment:
**thought:** the intersection between different keys seem a bit confusing

The following intersection:

```typescript
{
  [key in TKeys[number]]?: TypographyOptionsVariant
} & {
  [key in TKeys[0]]-?: TypographyOptionsVariant
}
```

Seems a bit hard to follow. Maybe you could add a comment to describe that the first entry in`TKeys` will result in a "required" `TypographyOptionsVariant`?

Reply:
Totally agree, comments coming up :)